### ISSUE TITLE
Fix auto update UX

### ISSUE THREAD

**Issue by benogle** [atom/atom#9164]
Our automatic update process is pretty bad. We release once a week, and every week over a half million people need to go through this process.

There are [errors](https://github.com/atom/atom/issues/7854), [bugs](https://github.com/atom/atom/issues/6482), folks want [better feedback](https://github.com/atom/atom/issues/8038), and the squirrel icon + release notes are really really confusing:
- https://github.com/atom/settings-view/issues/511
- https://github.com/atom/release-notes/pull/38
- https://github.com/atom/atom/issues/5692
- https://github.com/atom/atom/issues/6922
- https://github.com/atom/atom/issues/3757
- https://github.com/atom/atom/issues/8265
### What happens now
- Download an [old version](https://github.com/atom/atom/releases/tag/v1.0.18)
- Start it
- Notice there is a blue squirrel in the status bar, and it says 'view release notes'
- Click the squirrel
- See the release notes, note that they are showing notes for an old release
- Click the restart and update button
- Atom quits, notice that there is no feedback as to what is happening.
- It's still not back. What's up?
- Atom restarts again after 5 - 15 seconds
- Notice that the release notes are up in the reloaded window. They are still out of date, and still have the 'Restart and Update' button showing and enabled
- Notice that the squirrel icon is still showing, and still blue, with the same tooltip text
- I click the squirrel because it's still there, and not going away. It focuses the release notes. Ok.
- I'm thinking: is there another update??!? What is happening?
- I check the atom menu
  - Click the 'About' item, theres no information saying it's the latest version
  - I notice there is a version number, and a 'Check for update'
  - Nothing indicates I'm at the latest version
- On reload, squirrel goes away; still nothing tells me I'm at the latest version. But no squirrel, so I guess so?
### A 70% fix

tl;dr: fix the squirrel icon, remove the release notes package, use the about package for update status (similar to chrome).
- An update becomes available
- The blue squirrel shows, its tooltip says "Downloading update for version 1.0.20..."
- Clicking the squirrel shows the about page which shows
  - the current version
  - new version detected
  - some text indicating that the update is being downloaded
- When it's finished downloading:
  - the blue squirrel turns green, with a tooltip of "Version 1.0.20 is available"
  - the about page now shows that an update is available, and has a button to restart the editor
  - The menu item now takes you to the about page
- Clicking the green squirrel still shows / focuses the about page
- Clicking reload reloads the editor like today
- On reload the about page is shown.
- The about page indicates that you are up to date, and has a link to the release notes
- The squirrel is _gone_ after reload as the about page has the relevant information

Additionally, take a stab at the bugs:
- https://github.com/atom/atom/issues/7854
- https://github.com/atom/atom/issues/6482

Bonus:
- In the about page, indicate which release rail you are on. beta? stable? development?
### An extra 10%

Add some kind of feedback when atom is reloading. A small window with an infinite spinner would be sufficient. The goal is to indicate that it is working.
### That last 20%
- Add progress to the download in squirrel (cc @joshaber, is this possible today?)
  - Add progress to the menu item in %
  - Add progress to the squirrel icon
    - Could add a progress bar to the tooltip
    - Could make the blue squirrel "fill up" with green as it progresses
- Add an option to turn off automatic downloads in the about page to fix https://github.com/atom/atom/issues/1882

**Comment by joshaber** [comment]
>Add progress to the download in squirrel (cc @joshaber, is this possible today?)

Not currently :disappointed: Certainly something that could be added.

**Comment by benogle** [comment]
What's difficulty / time commitment on something like that?

**Comment by joshaber** [comment]
>Atom restarts again after 5 - 15 seconds

This is a problem for GitHub Desktop on Mac too. Every time we do an update, we get crashes reported because users try to relaunch the app themselves since the updater takes so long to restart the app.

**Comment by joshaber** [comment]
>What's difficulty / time commitment on something like that?

The work itself isn't too hard. The time consuming part is testing. Probably a few days worth of work.

**Comment by benogle** [comment]
:heart: for the info about the mac update issues and the difficulty level.

Would a UI for restart progress make sense in squirrel or should we subscribe to something and show our own window?

**Comment by joshaber** [comment]
>Would a UI for restart progress make sense in squirrel or should we subscribe to something and show our own window?

Restart progress or download progress? Restart progress would be tricky since the app is quit at that point.

**Comment by benogle** [comment]
> Restart progress

Yeah some kind of restart feedback. e.g. sketch has a little window with a progress bar when it's restarting after an update.

**Comment by joshaber** [comment]
Oh interesting. Squirrel *could* do that, though it's a non-trivial amount of effort.

It'd also be interesting to see *why* restarting takes so long. I suspect there's work we could do up front, before we relaunch, that would shorten the restarting window.

**Comment by 50Wliu** [comment]
I think download progress can be shown on Squirrel for Windows; I just haven't had enough time yet to dig into the code and try to implement that.  Though now that I've actually built Atom I might be able to make some progress soon :tm:.

also, /cc @paulcbetts and refs #8038

[All `auto-updater`-labeled issues](https://github.com/atom/atom/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Aauto-updater+-label%3Amore-information-needed+-label%3Aneeds-reproduction)
[All `installer`-labeled issues](https://github.com/atom/atom/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Ainstaller+-label%3Amore-information-needed+-label%3Aneeds-reproduction+)

**Comment by Zireael07** [comment]
The squirrel changing color would be a brilliant quick change.

**Comment by paulcbetts** [comment]
Squirrel for Windows reports download progress already, in a very node.js friendly way, you just need to wire it up. That being said though, most of the work in Squirrel on Windows isn't *downloading*, it's reconstituting the full package from the delta. Ideally the download step of your download will be super-ass short

**Comment by maxbrunsfeld** [comment]
Refs https://github.com/atom/atom/issues/8911

**Comment by ethanluoyc** [comment]
@benogle Are we trying to make something like this?

![about](https://cloud.githubusercontent.com/assets/6040760/11181389/b1835c68-8c9c-11e5-9d42-032b60da87c5.gif)

I noticed that if we want to check for updates we may need to use the auto-update manager from the core. The about page, however, it organized as a separate package

**Comment by benogle** [comment]
> we will just open up the about page from the squirrel

Yes. The release notes package will go away. It has a lot of problems, and it's just duplicating the atom/atom/releases page anyway, so we can just link to that page.

**Comment by ethanluoyc** [comment]
@benogle What are the best ways I can check for updates? currently the release-notes package uses caching to store versions and use that to check against, but I think something simpler can be done?

**Comment by ethanluoyc** [comment]
started working on this and first commit already submitted to https://github.com/atom/about/pull/5, but I am quite busy for the next couple of weeks so will not be able to finish this really soon. But I guess we can discuss on possible improvements based on this commit?

**Comment by YuriSolovyov** [comment]
It would be great if linux version could at least check for a newer version.

**Comment by ethanluoyc** [comment]
Currently the implementation is the appearance of the squirrel if the editor emits an update-available event. Clicking on the squirrel will open the Atom release page where you can download the newest update. But I think that we would prefer an auto download and update mechanism like we had now?

**Comment by YuriSolovyov** [comment]
@ethanluoyc I think for linux it should be delivered via system's auto update mechanism based on repositories.
My comment was only about "hey! there is an update available!" notification.

**Comment by mnquintana** [comment]
@YuriSolovyov @ethanluoyc Yup, that's definitely something I've wanted to add to about (it's currently being tracked in https://github.com/atom/about/issues/1)

**Comment by Alhadis** [comment]
Are there any plans to do away with the usage of the US-only date order for timestamping releases? Kinda selfish seeing MM/DD/YYYY for an internationally-distributed product.

Whole world should just shake hands and use ISO's format, period.

**Comment by Neosublimation** [comment]
Actually, Atom appears to really not restart at all after an update for me. The first few updates it did pretty well, but at some point the release notes stopped updating (they're still showing the notes for 1.1.0) and around that point Atom stopped auto-restarting after an upgrade. That or my computer is really much slower than I thought. I remember not manually restarting it after clicking the update button for half an our or so, because I didn't need it at that time so I thought it can take its time, and it didn't start again within this timeframe (yet it appears to have updated correctly).

I just updated from 1.3 to 1.3.1 and it didn't restart on its own and I couldn't find an active process once it was shut down.

But it's really confusing. If there wasn't a version info in the help menu I'd think I'm stuck at 1.1.0 for some reason.

**Comment by Ivoz** [comment]
I had **no** idea to look for or notice a squirrel in the bottom right hand corner before checking this issue and reading of it. I'm guessing a major percentage of the average user will never notice it either, in case its intended purpose is being noticed.

I just updated 1.0.19 -> 1.3.2 and clicking the "Restart and Install Update" menu option worked as described, on x64 Windows 10.

I would love to see some kind of progress % update UI somewhere for updating, doesn't have to be super accurate, update every second or 5, or show only every 10% made, etc.

**Comment by Zireael07** [comment]
Release notes has been removed and I think the blue squirrel went away, too. 

No indication that atom is reloading nor any sort of indication that I'm now running the latest. And no % info on download progress, either.

Are these going to happen?

**Comment by maxbrunsfeld** [comment]
>  I think the blue squirrel went away, too.

The squirrel should still be present. Are you getting an update and not seeing it?

**Comment by judahstevenson** [comment]
I'm on OS 10.10 and I do see the blue squirrel still. Clicking on it took me to release notes page, but it was not clear that there was an update available until I read through some posts and checked the Atom menu where I saw "Restart and Install Update". That worked, but there was no indication afterwards that the update had succeeded, i.e. no confirmation message, feedback, etc. But when I checked the Atom menu again and selected "Check for Update" it confirmed that I was on the latest version. Majority of my confusion I believe was trying to figure out the new update procedure since I am still accustomed to the old procedure. But I still feel that this was an improvement, however minor.

**Comment by benogle** [comment]
Thanks for the feedback. This is definitely something we are working on at the moment.

**Comment by jokeyrhyme** [comment]
@judahstevenson this is pretty much the Chrome update experience: no notifications, it just works behind the scenes, and you have the manually check

@benogle has it been determined that Atom should strive to have a different / better update experience than Chrome?

**Comment by benogle** [comment]
Basically we're adding some things to the about page to be more like chrome. e.g. 

![f5a343d6-c5cc-11e5-8ca7-46a320c16c86](https://cloud.githubusercontent.com/assets/69169/12658919/b4a09fde-c5c0-11e5-94b2-1a362f6378fe.png)

Plus some feedback when it's downloading, when it's already up to date, etc. You can deal with it just like chrome and restart at your leisure, or you can pay attention to the squirrel and do a similar update flow from the about page.

**Comment by benogle** [comment]
Just an update. This is coming along in:

* Foundational work: https://github.com/atom/atom/pull/10675
* UI work: https://github.com/atom/about/pull/12

**Comment by Stanzilla** [comment]
This is now super weird: https://github.com/atom/about/issues/24

**Comment by Stanzilla** [comment]
Related: https://github.com/atom/atom/issues/4696

**Comment by jhabdas** [comment]
Apologies if this is not more constructive. But I found myself for 15 minutes staring at this:

<img width="514" alt="screen shot 2017-06-05 at 5 08 36 pm" src="https://cloud.githubusercontent.com/assets/440298/26778232/fbc6449a-4a12-11e7-9a8e-c64c44447b7f.png">

And praying that, eventually, I would see this (which I eventually did):

<img width="512" alt="screen shot 2017-06-05 at 5 17 05 pm" src="https://cloud.githubusercontent.com/assets/440298/26778231/fbc4b882-4a12-11e7-837d-1b7aa9d735b5.png">

Is this the place to discuss adding a % complete or KB of TOTAL for Atom?

**Comment by 50Wliu** [comment]
@jhabdas Based on the conversation in this issue, it would appear that Squirrel.Mac doesn't report download percentages.  Squirrel.Windows does, but no one has had time to investigate how to hook that up with Atom.
(Squirrel is the update framework Atom uses, though the Windows version and Mac version are maintained by different people and work differently)

**Comment by jhabdas** [comment]
> Squirrel.Mac doesn't report download percentages. Squirrel.Windows does

@50Wliu Thanks that's super helpful. Sorry I didn't take as much time as I should've parsing the thread. Seems like the next step is...

**Comment by movingobjects** [comment]
It's been a while since there's been any update on this—still find this very confusing every single update. (I'm on macOS 10.13)

It'd be great to have more feedback on what is going on. Clicking the `Restart and Install` button only occasionally seems to actually restart the app itself. More often, it closes all windows but Atom remains open in the dock. I then wait and wonder whether a new version is downloading/installing or Atom just didn't succeed in restarting itself. After 5-10 minutes, I'll give up and quit and re-open Atom manually—after which, sometimes I end up with the new version, and sometimes I don't and instead see the same `Restart and Install` button sitting there like I never clicked it. After a few rounds of this, I will eventually find myself with the new version, but after a year of updates, I honestly still don't know what I'm supposed to expect or do to update correctly.

**Comment by lee-dohm** [comment]
Thanks very much for the feedback.

Unfortunately, because of the disparity in the APIs between the implementations of Squirrel, we're not going to take on adding a progress indicator for the automatic updates at this time. Because of this, and because since this issue has been opened, we've made a bunch of progress on improving the auto-update UX, we're going to close this issue.

If people have other feedback for fixing the auto-update experience beyond adding a progress indicator, please open a new issue for it.

Thanks everyone for your help!
